### PR TITLE
Clean up some scheduler APIs

### PR DESCRIPTION
**Pull request by lrh2000**
### The PR description below is outdated, please see https://github.com/asterinas/asterinas/pull/1472#issuecomment-2444496855 and https://github.com/asterinas/asterinas/pull/1472#issuecomment-2427078063 directly.

I'm not sure why we introduce `ReschedAction::Retry`:
https://github.com/asterinas/asterinas/blob/2f511069eecf521661aa4d8ee3020863be75b543/ostd/src/task/scheduler/mod.rs#L236-L244

Looking at the code, `ReschedAction::Retry` is clearly redundant because we can loop over `reschedule()`:
https://github.com/asterinas/asterinas/blob/2f511069eecf521661aa4d8ee3020863be75b543/ostd/src/task/scheduler/mod.rs#L209-L230

In some cases the outer loop already exists, so the inner loop doesn't make much sense to me:
https://github.com/asterinas/asterinas/blob/2f511069eecf521661aa4d8ee3020863be75b543/ostd/src/sync/wait.rs#L270-L272

The outer loop in `Waiter` checks `has_woken` on every iteration (as shown above). However, in the inner loop we can only check the flag on the first iteration. Ideally we should check the flag on every iteration, but we cannot do that because we cannot return `DoNothing` on the second or later iteration unless `pick_next_current` returns the original current.
https://github.com/asterinas/asterinas/blob/2f511069eecf521661aa4d8ee3020863be75b543/ostd/src/task/scheduler/mod.rs#L113-L134

To solve the problem, I think we should merge `dequeue_current` and `pick_next_current` into one method. The proposed API is:
```rust
/// The state of the current task at the time of scheduling.
#[derive(PartialEq, Copy, Clone)]
pub enum CurrentState {
    /// The current task is still runnable, so it will be kept in the runqueue.
    Runnable,
    /// The current task needs to go to sleep, so it will leave the runqueue.
    NeedSleep,
}

pub trait LocalRunQueue<T = Task> {
    /// Picks the next runnable task.
    ///
    /// This method returns the selected next runnable task. This task will replace the current
    /// runnable task after the method returns. If there is no candidate for such a task, this
    /// method will do nothing but return `None`.
    ///
    /// After the next runnable task is selected, `current_state` will decide whether or not to
    /// keep the current runnable task in the run queue. See [`CurrentState`] for details.
    fn pick_next(&mut self, current_state: CurrentState) -> Option<&Arc<T>>;
}
```

### CODE REVIEW

**Comment by lrh2000** [comment]
NOTE: This PR is _not_ related to https://github.com/asterinas/asterinas/issues/1471.

However, I _did_ add a FIXME for that issue:
```rust
    // FIXME: At this point, we need to prevent the current task from being scheduled on another
    // CPU core. However, we currently have no way to ensure this. This is a soundness hole and
    // should be fixed. See <https://github.com/asterinas/asterinas/issues/1471> for details.
```

**Comment by tatetian** [comment]
@js2xxx @jellllly420 Could you share your opinions about the proposed changes to the scheduler interface?

**Comment by js2xxx** [comment]
> Taking a closer look at park_current, if the first iteration does dequeue_current and returns ReschedAction::Retry and the second iteration returns ReschedAction::DoNothing, we will leave the scheduler with an empty current field and run the rest of the code, which is extremely odd.

`pick_next_current` will set `current` to the picked task, so it won't be left empty.

**Comment by lrh2000** [comment]
> pick_next_current will set current to the picked task, so it won't be left empty.

Oh, sorry. That's a mistake. Withdraw this claim (I've edited the PR description).

Well, it's not a bug, but I think as a cleanup PR the changes still make sense. I don't really like the previous `Retry` design because:
1. It is redundant, since we can loop over `reschedule()`. Removing it can make our codebase cleaner (as shown in this PR) while maintaining all functionality.
2. Specifically for `unpark_current`, the outer loop in `Waiter` checks `has_woken`. In the inner loop we can only check the flag on the first iteration. Ideally we should check the flag on every iteration, but we cannot do that because we cannot return `DoNothing` on the second or later iteration _unless `pick_next_current` returns the original current_.

**Comment by js2xxx** [comment]
> unless pick_next_current returns the original current.

In fact, it is not expected to happen because a task cannot be in the run queue to be picked as the next current task while running already. In other words, `DoNothing` is returned only when `has_woken` is set. The `Arc::ptr_eq` is only a defense to state spill.

> make our codebase cleaner while maintaining all functionality.

As you said in #1471, we cannot expect user-defined schedulers to behave correctly. However, you leave `CurrentState` to the scheduler and expect it to handle it as expected, which is a new case of state spill. This is not a way to make our codebase cleaner while maintaining all functionality.

**Comment by lrh2000** [comment]
> In fact, it is not expected to happen because a task cannot be in the run queue to be picked as the next current task while running already. In other words, `DoNothing` is returned only when `has_woken` is set. The `Arc::ptr_eq` is only a defense to state spill.

I don't think so. This can happen due to race conditions.

The task can be woken up after checking the `is_woken` flag (note that the check only happens at the beginning of the first iteration). After `deuque_current` but before `switch_to_task`, the current task can be put back into the run queue.

From the perspective of #1471, I also want to _make sure that the time between `dequeue_current` and `switch_to_task` is as small as possible_. The new API does this, but with the old API there is a loop between `dequeue_current` and `switch_to_task` that can run for a while.
 - This is because one potential strategy to fix #1471 is: if we find that a task is scheduled on the second CPU, but the task is still running on the first CPU, we can spin until the task is switched out on the first CPU. This strategy will fix the soundness hole, and won't hurt efficiency if the correctness of the scheduler implementation ensures that the spin time is always short.

> As you said in #1471, we cannot expect user-defined schedulers to behave correctly. However, you leave `CurrentState` to the scheduler and expect it to handle it as expected, which is a new case of state spill. This is not a way to make our codebase cleaner while maintaining all functionality.
That sounds even more wired to me. My feelings are exactly the opposite of what you're trying to argue.

In the old APIs, we're actually requiring custom schedulers to correctly implement two APIs, `dequeue_current` and `pick_next_current`. We require custom schedulers to correctly implement not only the two APIs themselves, _but also the way they interact_.

Now, I don't think it's obvious how these two APIs interact. For example, we call `dequeue_current` and repeat the call to `pick_next_current` until `pick_next_current` returns `Some(_)` to ensure that `current` doesn't remain `None`. This means that we're breaking the invariant of the custom scheduler (and then restoring the invariant). 

```rust
impl Scheduler {
    // Breaks the invariant: Now `current` is `None`.
    fn dequeue_current(&mut self) { /* .. */ }

    // May restore the invariant: Now `current` may become `Some(_)`.
    fn pick_next_current(&mut self) { /* ... */ }

    // Question: Can this method be called while the invariant is broken?
    //
    // We know this won't happen, but the only way to find out is to read the OSTD code carefully.
    fn enqueue(&mut self) { /* ... */ }
}
```

However, a significant advantage of the new APIs is that we will know that `current` cannot be empty (which is an invariant of the scheduler itself) _just by looking at the scheduler's code_. Before, we had to check both the scheduler code and the OSTD code.

```rust
impl Scheduler {
    // Never break the invariant: If `current` is `Some(_)` before, `current` will be `Some(_)` after.
    fn pick_next(&mut self, current_state: CurrentState) -> Option<&Arc<T>>;
}
```

 - This invariant is an invariant of the custom scheduler. OSTD itself can never guarantee that it is correctly guaranteed, so the best we can do is design good APIs to allow the custom scheduler to guarantee it more flexibly (instead of relying facts about the OSTD implementation).

> However, you leave `CurrentState` to the scheduler and expect it to handle it as expected

So are you saying that if we keep the old APIs, `dequeue_current`/`pick_next_current` cannot be incorrectly implemented by an evil or malfunctioning scheduler? To be honest, I'm really confused about this..

**Comment by js2xxx** [comment]
In fact, I think this PR should be postponed until #1192 is merged, which will fully utilize the current scheduling API and expose most of its bugs/design flaws. Then the incoming fixes should be more practical instead of arguing over superficial design choices.

**Comment by jellllly420** [comment]
I see two major disagreements here and I'd like to explain how I think about these one by one:
- whether `ReschedAction::Retry` is redundant

  It really depends on how one thinks about the responsibility sharing between scheduling subsystem and its users.
  
  Let's take your `do_wait` as example. The `do_wait -> park_current -> reschedule` chain is somewhat like a try-catch procedure. We try to find a new task to replace the current sleeping task and the "finding" will emit an exception if there are no candidates or the sleeping task has been woken (but once we find a new task, it will always succeed no matter what status the sleeping task is in). 

  Under current design, the `SwitchTo` is for the successful case and the `Retry` and `DoNothing` is for the failed case. Since `reschedule` is a `scheduler` moudle-internal API, we can say that the `scheduler` module controls the overall try-catch procedure and provides limited number of patterns (e.g. `park_current`) of the above try-catch procedure, which should be just enough for OSTD-internal use and don't need a "outer loop" like that one in `do_wait` (I re-examined the `do_wait` logic and found you can simply use `if` to replace `while` there).

  If we remove the `Retry` and switch to "outer loops" around `reschedule`, it means we'd like the users of our scheduling API's (e.g. `park_current`, not the `Scheduler` trait) to take some responsibility (handle the cases that previously handled by `Retry`) in the above try-catch procedure and involve some scheduling-related logics.

  From my prospective, I'd like the API is just an API and don't let the scheduling logic escape to its users (e.g. `do_wait`) and therefore I prefer keeping the `Retry`.

- which level of safety OSTD API's should gurantee
  
  I think this PR has nothing to do with safety or evil scheduler developers. I agree with @lrh2000; neither the current API's or the proposed API's (in fact, I think, no API's) can ensure "correct" behaviors under "incorrect" scheduler implementaions. 

  Actually I think what OSTD can do is to provide a correct mechanism since we cannot control what policy our users may implement. As to the policy, I think there have to exist some conventions or specifications between OSTD and its users. That's why I think #1471 is a real soundness problem of OSTD while this PR is not.

**Comment by jellllly420** [comment]
@lrh2000 Here are some of my thoughts about your arguments:

> 1. It is redundant, since we can loop over `reschedule()`. Removing it can make our codebase cleaner (as shown in this PR) while maintaining all functionality.

As I said above, the proposed changes surely decrease LoC and make some logic more clear, but in the price of the scheduling logic escaping to the users.

> 2. Specifically for `unpark_current`, the outer loop in `Waiter` checks `has_woken`. In the inner loop we can only check the flag on the first iteration. Ideally we should check the flag on every iteration, but we cannot do that because we cannot return `DoNothing` on the second or later iteration unless `pick_next_current` returns the original `current`.

You are right about this, "we cannot return `DoNothing` on the second or later iteration unless `pick_next_current` returns the original `current`". But I don't quite agree that "Ideally we should check the flag on every iteration". We surely can avoid a context-switch in an early-wakeup scenario (on the second or later iteration) if we check `has_woken` on every iteration but I'm not very sure if the use case is concrete enough. Since there is an idle thread per CPU, the iteration typically should terminate on the first round and the real time window is the check at the beginning of the first round and the `SwitchTo` on the first round. The proposed change cannot handle the early-wakeup in this time window either (please correct me if i'm wrong).

**Comment by lrh2000** [comment]
> In fact, I think this PR should be postponed until #1192 is merged, which will fully utilize the current scheduling API and expose most of its bugs/design flaws. Then the incoming fixes should be more practical instead of arguing over superficial design choices.

@js2xxx  If you claim that PR significantly affects the decision-making process in this PR (or just significantly conflicts with this PR), I don't mind postponing the major changes to this PR to future work. 

> It really depends on how one thinks about the responsibility sharing between scheduling subsystem and its users.

@jellllly420 Yes, this is a valid argument.

My personal feeling is that the scheduling APIs and their users can hardly be completely decoupled. After all, `park_current` must take the `has_woken` flag (or a closure that takes the flag), and the waiter cannot be implemented correctly if it does not understand how `unpark_current` is implemented (otherwise it will surely encounter some race conditions).

So I've given this argument a slightly lower rank. But I admit it's valid.

> (I re-examined the do_wait logic and found you can simply use if to replace while there).

No, this is not correct.

Waiters can be nested. We can create two `Waiter`s/`Waker`s for one task at the same time. The problem is that `unpark_current` _cannot_ guarantee that there will be no spurious wake events (it may just wake up another waker but spuriously wake up the current task). You can have a look at the `waiter_wake_reorder` unit test for a concrete example.

> From my prospective, I'd like the API is just an API and don't let the scheduling logic escape to its users (e.g. do_wait) and therefore I prefer keeping the Retry.

> As I said above, the proposed changes surely decrease LoC and make some logic more clear, but in the price of the scheduling logic escaping to the users.

We have to explicitly allow `park_current` to have spurious wake events (see above) and the caller must have some mechanisms to deal with spurious wake events. So, instead of "letting the scheduling logic escape to its users", what you only need to agree that "missing available tasks is a kind of spurious wake event"..

**Comment by lrh2000** [comment]
> As I said above, the proposed changes surely decrease LoC and make some logic more clear, but in the price of the scheduling logic escaping to the users.

Besides the points above, another option is to try this (I think it does a slightly better job from the point of view of splitting responsibility between the scheduling subsystem and its users):
```diff
diff --git a/ostd/src/task/scheduler/mod.rs b/ostd/src/task/scheduler/mod.rs
index fb2a83f0..e278d76d 100644
--- a/ostd/src/task/scheduler/mod.rs
+++ b/ostd/src/task/scheduler/mod.rs
@@ -114,11 +114,12 @@ pub(crate) fn might_preempt() {
     yield_now();
 }
 
-/// Blocks the current task unless `has_woken` is `true`.
+/// Blocks the current task until `has_woken` is `true`.
 pub(crate) fn park_current<F>(mut has_woken: F)
 where
     F: FnMut() -> bool,
 {
+    while !has_woken() {
         reschedule(|local_rq: &mut dyn LocalRunQueue| {
             if has_woken() {
                 return None;
@@ -136,6 +137,7 @@ where
             local_rq.update_current(UpdateFlags::Wait);
             local_rq.pick_next(CurrentState::NeedSleep)
         });
+    }
 }
 
 /// Unblocks a target task.
```

**Comment by lrh2000** [comment]
**Summary and actions:**

First of all, thanks @js2xxx and @jellllly420 for all the conversation we've had.

Following our discussion, and to make things actionable, I've broken the commits into more atomic ones, and ordered them by their level of controversy. If any of them are deemed undesirable (or just need to be postponed), I'll be happy to remove them from this PR.

In summary, I think:
 - The first two commits contain changes that are non-controversial (hopefully?).
 - [The third commit](https://github.com/asterinas/asterinas/pull/1472/commits/95ea1ddeec3be78b865950805dfbb54d3b3d8ffa) combines `dequeue_current` and `pick_next_current` into one method called `pick_next`. Although @js2xxx expressed his opposition (https://github.com/asterinas/asterinas/pull/1472#issuecomment-2425455407), I do not understand his point. From my perspective, this brings benefits and has no drawbacks (https://github.com/asterinas/asterinas/pull/1472#issuecomment-2425580879).
 - [The fourth commit](https://github.com/asterinas/asterinas/pull/1472/commits/6a5663a1357268948ee8ad984c89ff09062f3181) removes ReshedAction::Retry. Both @js2xxx and @jellllly420 objected to this, where I admit that some points are valid (e.g. https://github.com/asterinas/asterinas/pull/1472#issuecomment-2425781937). However, I have also detailed my argument (https://github.com/asterinas/asterinas/pull/1472#issuecomment-2426003219).

**Comment by lrh2000** [comment]
**Update:** I discussed the third and fourth commits with @tatetian at last Friday's meeting. @tatetian raises a concern that makes a lot of sense to me: After we run `dequeue_current`, ticks will not count for the dequeued task, but by merging `dequeue_current` and `pick_next_current`, ticks will still count while the current task wants to sleep but we cannot find the next task.

After careful consideration, here are my conclusions and actions:
 - The [fourth commit](https://github.com/asterinas/asterinas/commit/6a5663a1357268948ee8ad984c89ff09062f3181) will make the above problem literally unfixable, so it will be removed from this PR.
 - With the [third commit](https://github.com/asterinas/asterinas/commit/95ea1ddeec3be78b865950805dfbb54d3b3d8ffa), we can still find a way to fix the above problem, but at the cost of adding more APIs to the `LocalRunQueue` trait. Therefore, this will not be an API cleanup (as it can be seen as adding complexity to the final APIs instead). I will _postpone_ this commit.
   - I say _postpone_ because it may still help resolve #1471, as I mentioned in https://github.com/asterinas/asterinas/pull/1472#issuecomment-2425580879. However, it does not make sense as a standalone API cleanup. An RFC describing the possible ways to fix #1471 has not yet been written, but I will do so later. At that time, I will discuss the related matters again.

I also checked @js2xxx's #1472, and this PR should now have no significant conflict (just a one-line conflict in `ostd/src/task/scheduler/mod.rs`, which should be totally fine, I think).

File: ostd/src/task/scheduler/mod.rs [hunk 1]
```diff
@@ -114,27 +112,45 @@ pub(crate) fn might_preempt() {
...
 
-/// Blocks the current task unless `has_woken` is `true`.
-pub(crate) fn park_current(has_woken: &AtomicBool) {
+/// Blocks the current task unless `has_woken()` returns `true`.
+///
+/// Note that this method may return due to spurious wake events. It's the caller's responsibility
+/// to detect them (if necessary).
+pub(crate) fn park_current<F>(mut has_woken: F)
+where
+    F: FnMut() -> bool,
```
Comment (tatetian) [inline ostd/src/task/scheduler/mod.rs:121]: Taking a mutable closure `FnMut` looks weird for a closure that is supposed to check if a condition turns true.

Comment (lrh2000) [reply ostd/src/task/scheduler/mod.rs]: Oh, that's a good catch! Thanks!

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**